// @teambit/design.ui.error-page@0.0.364/dist/contact-footer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContactIcons = void 0;
const react_1 = __importDefault(require("react"));
const base_react_navigation_link_1 = require("@teambit/base-react.navigation.link");
const community_constants_links_1 = require("@teambit/community.constants.links");
const styles = {
    link: {
        textDecoration: 'none',
        lineHeight: 'inherit',
        color: 'unset',
    },
    logo: {
        width: '24px',
        height: '24px',
        margin: '0 13px',
    },
};
// TODO: get the docs domain from the community aspect and pass it here as a prop
function ContactIcons(props) {
    return (react_1.default.createElement("div", Object.assign({ style: Object.assign(Object.assign({}, styles.iconLine), props.style) }, props),
        react_1.default.createElement(base_react_navigation_link_1.Link, { external: true, style: styles.link, href: community_constants_links_1.links.slack },
            react_1.default.createElement("img", { alt: "slack", style: styles.logo, src: "https://static.bit.dev/harmony/slack-round-icon.svg" })),
        react_1.default.createElement(base_react_navigation_link_1.Link, { external: true, style: styles.link, href: community_constants_links_1.links.github },
            react_1.default.createElement("img", { alt: "github", style: styles.logo, src: "https://static.bit.dev/harmony/github.svg" })),
        react_1.default.createElement(base_react_navigation_link_1.Link, { external: true, style: styles.link, href: community_constants_links_1.links.docs },
            react_1.default.createElement("img", { alt: "bit docs", style: styles.logo, src: "https://static.bit.dev/bit-logo.svg" }))));
}
exports.ContactIcons = ContactIcons;
