// @teambit/design.ui.error-page@0.0.364/dist/error-page.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorPage = void 0;
const react_1 = __importDefault(require("react"));
const styles = {
    errorPage: {
        display: 'flex',
        flexDirection: 'column',
        justifyContent: 'center',
        alignItems: 'center',
        margin: 'auto',
        padding: '24px',
        textAlign: 'center',
    },
    title: {
        fontWeight: 'bold',
        margin: '0 0 24px 0',
    },
    img: {
        maxWidth: '700px',
        marginBottom: '24px',
        width: '100%',
    },
};
const inlineStyles = `
.bit-error-page-title { font-size: 36px; }
@media screen and (max-width: 480px) { .bit-error-page-title { font-size: 24px; } }
`;
/**
 * A component that shows an error page according to the error code
 */
function ErrorPage(_a) {
    var { code, title, style, children } = _a, rest = __rest(_a, ["code", "title", "style", "children"]);
    return (react_1.default.createElement("div", Object.assign({}, rest, { style: Object.assign(Object.assign({}, styles.errorPage), style) }),
        react_1.default.createElement("style", null, inlineStyles),
        react_1.default.createElement("h1", { style: styles.title, className: "bit-error-page-title" }, title),
        react_1.default.createElement("img", { alt: "error-image", style: styles.img, src: `https://static.bit.dev/harmony/${code}.svg` }),
        children));
}
exports.ErrorPage = ErrorPage;
