// @teambit/design.ui.pages.not-found@0.0.366/dist/not-found-page.docs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const documenter_ui_section_1 = require("@teambit/documenter.ui.section");
const documenter_theme_theme_compositions_1 = require("@teambit/documenter.theme.theme-compositions");
const documenter_ui_separator_1 = require("@teambit/documenter.ui.separator");
const not_found_page_1 = require("./not-found-page");
function Overview() {
    return (react_1.default.createElement(documenter_theme_theme_compositions_1.ThemeCompositions, null,
        react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(documenter_ui_section_1.Section, null, "A page component that displays a 404 error message."),
            react_1.default.createElement(documenter_ui_separator_1.Separator, null))));
}
exports.default = Overview;
Overview.abstract = 'A 404 error page.';
Overview.labels = ['react', 'typescript', '404', 'error', 'page'];
Overview.examples = [
    {
        scope: {
            NotFoundPage: not_found_page_1.NotFoundPage,
        },
        title: '404 page',
        description: '404 error message',
        jsx: react_1.default.createElement(not_found_page_1.NotFoundPage, null),
    },
];
