// @teambit/react.generator.react-templates@1.0.10/dist/app-template/app-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactAppTemplate = void 0;
const index_file_1 = require("./files/index-file");
const app_1 = require("./files/app");
const react_app_1 = require("./files/react-app");
const composition_1 = require("./files/composition");
const app_root_1 = require("./files/app-root");
class ReactAppTemplate {
    constructor(name = 'react-app', description = 'a template for generating react app components', hidden = false, env) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, composition_1.compositionFile)(context),
            (0, app_1.AppFile)(context),
            (0, react_app_1.reactAppFile)(context),
            (0, app_root_1.AppRootFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new ReactAppTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.ReactAppTemplate = ReactAppTemplate;
