// @prerenderer/webpack-plugin@5.3.8/dist/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = exports.defaultOptions = void 0;
exports.defaultOptions = {
    indexPath: 'index.html',
    routes: ['/'],
};
exports.schema = {
    type: 'object',
    properties: {
        entryPath: {
            description: 'The entry index.html file to use',
            type: 'string',
            nullable: true,
        },
        fallback: {
            oneOf: [
                { type: 'string' },
                { type: 'boolean' },
            ],
            description: 'Generate a fallback file if the asset already exists',
        },
        routes: {
            description: 'A list of routes to pre-render',
            type: 'array',
            nullable: true,
            items: {
                description: 'The path of the route',
                type: 'string',
            },
        },
        urlModifier: {
            instanceof: 'Function',
            description: 'Hook to be able to modify the url to retrieve the compiled asset',
        },
    },
};
