// @teambit/react.generator.react-templates@1.0.10/dist/component-template/component-template.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReactComponentTemplate = void 0;
const index_file_1 = require("./files/index-file");
const component_1 = require("./files/component");
const test_1 = require("./files/test");
const composition_1 = require("./files/composition");
class ReactComponentTemplate {
    constructor(name = 'react', description = 'a template for react components', hidden = false, env) {
        this.name = name;
        this.description = description;
        this.hidden = hidden;
        this.env = env;
    }
    generateFiles(context) {
        return [
            (0, index_file_1.indexFile)(context),
            (0, composition_1.compositionFile)(context),
            (0, component_1.componentFile)(context),
            (0, test_1.testFile)(context),
        ];
    }
    static from(options = {}) {
        return () => new ReactComponentTemplate(options.name, options.description, options.hidden, options.env);
    }
}
exports.ReactComponentTemplate = ReactComponentTemplate;
