// @prerenderer/renderer-jsdom@1.1.7/dist/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schema = exports.defaultOptions = void 0;
exports.defaultOptions = {
    injectProperty: '__PRERENDER_INJECTED',
    maxConcurrentRoutes: 0,
    skipThirdPartyRequests: false,
    timeout: 1000 * 30, // 30sec timeout default
};
exports.schema = {
    type: 'object',
    additionalProperties: true,
    properties: {
        maxConcurrentRoutes: {
            type: 'number',
            nullable: true,
        },
        renderAfterDocumentEvent: {
            type: 'string',
            description: 'The name of the event that should trigger the rendering of the page',
            nullable: true,
        },
        renderAfterElementExists: {
            type: 'string',
            description: 'Wait until this selector is found on the page',
            nullable: true,
        },
        renderAfterTime: {
            type: 'number',
            description: 'Time to wait for in ms before rendering the page',
            nullable: true,
        },
        timeout: {
            type: 'number',
            description: 'The time in ms after which we should stop waiting and throw an error',
            nullable: true,
        },
        injectProperty: {
            type: 'string',
            description: 'The key of the injected value into window',
            nullable: true,
        },
        JSDOMOptions: {
            type: 'object',
            description: 'Additional options for JSDOM',
            additionalProperties: true,
            nullable: true,
        },
    },
};
