// @teambit/react.react-env@1.0.34/dist/react-env.bit-env.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ReactEnv = void 0;
function _generator() {
  const data = require("@teambit/generator");
  _generator = function () {
    return data;
  };
  return data;
}
function _reactGenerator() {
  const data = require("@teambit/react.generator.react-templates");
  _reactGenerator = function () {
    return data;
  };
  return data;
}
function _reactGenerator2() {
  const data = require("@teambit/react.generator.react-starters");
  _reactGenerator2 = function () {
    return data;
  };
  return data;
}
function _reactApps() {
  const data = require("@teambit/react.apps.react-app-types");
  _reactApps = function () {
    return data;
  };
  return data;
}
function _typescript() {
  const data = require("@teambit/typescript");
  _typescript = function () {
    return data;
  };
  return data;
}
function _preview() {
  const data = require("@teambit/preview.react-preview");
  _preview = function () {
    return data;
  };
  return data;
}
function _typescript2() {
  const data = require("@teambit/typescript.typescript-compiler");
  _typescript2 = function () {
    return data;
  };
  return data;
}
function _defender() {
  const data = require("@teambit/defender.eslint-linter");
  _defender = function () {
    return data;
  };
  return data;
}
function _eslint() {
  const data = require("eslint");
  _eslint = function () {
    return data;
  };
  return data;
}
function _application() {
  const data = require("@teambit/application");
  _application = function () {
    return data;
  };
  return data;
}
function _defender2() {
  const data = require("@teambit/defender.jest-tester");
  _defender2 = function () {
    return data;
  };
  return data;
}
function _defender3() {
  const data = require("@teambit/defender.prettier-formatter");
  _defender3 = function () {
    return data;
  };
  return data;
}
function _typescript3() {
  const data = _interopRequireDefault(require("typescript"));
  _typescript3 = function () {
    return data;
  };
  return data;
}
function _builder() {
  const data = require("@teambit/builder");
  _builder = function () {
    return data;
  };
  return data;
}
function _workspaceConfigFiles() {
  const data = require("@teambit/workspace-config-files");
  _workspaceConfigFiles = function () {
    return data;
  };
  return data;
}
function _pkg() {
  const data = require("@teambit/pkg");
  _pkg = function () {
    return data;
  };
  return data;
}
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == typeof i ? i : String(i); }
function _toPrimitive(t, r) { if ("object" != typeof t || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != typeof i) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
// import webpackTransformer from './config/webpack.config';

class ReactEnv {
  constructor() {
    /**
     * name of the environment. used for friendly mentions across bit.
     */
    _defineProperty(this, "name", 'react');
    /**
     * icon for the env. use this to build a more friendly env.
     * uses react by default.
     */
    _defineProperty(this, "icon", 'https://static.bit.dev/extensions-icons/react.svg');
    _defineProperty(this, "tsconfigPath", require.resolve('./config/tsconfig.json'));
    _defineProperty(this, "tsTypesPath", './types');
    _defineProperty(this, "jestConfigPath", require.resolve('./config/jest.config'));
    _defineProperty(this, "eslintConfigPath", require.resolve('./config/eslintrc.js'));
    _defineProperty(this, "eslintExtensions", ['.ts', '.tsx', '.js', '.jsx', '.mjs']);
    _defineProperty(this, "prettierConfigPath", require.resolve('./config/prettier.config.js'));
    _defineProperty(this, "prettierExtensions", ['.js', '.jsx', '.ts', '.tsx', '.mjs', '.cjs', '.json', '.css', '.scss', '.md', '.mdx', '.html', '.yml', '.yaml']);
    _defineProperty(this, "previewMounter", require.resolve('./preview/mounter'));
    _defineProperty(this, "previewDocsTemplate", require.resolve('./preview/docs'));
    /**
     * Default npm ignore paths.
     * Will ignore the "artifacts" directory by default.
     */
    _defineProperty(this, "npmIgnore", [`${_builder().CAPSULE_ARTIFACTS_DIR}/`]);
    /**
     * Default host dependencies for the react preview.
     */
    _defineProperty(this, "hostDependencies", _preview().ReactPreview.hostDependencies);
    /**
     * Default package.json modifications.
     */
    _defineProperty(this, "packageJson", {
      main: 'dist/{main}.js',
      types: '{main}.ts'
      // type: 'module',
    });
  }
  /**
   * return an instance of a Compiler. use components like typescript-compiler (teambit.typescript/typescript-compiler)
   * or our babel-compiler (teambit.compilation/babel-compiler).
   */
  // TODO: make sure we only run compiler once @david.
  compiler() {
    // const babelConfig = this.react.getBabelDefaultConfig();
    // const babelCompiler = new BabelCompiler(babelConfig);
    return _typescript2().TypescriptCompiler.from({
      tsconfig: this.tsconfigPath,
      // types: resolveTypes(__dirname, [this.tsTypesPath]),
      typescript: _typescript3().default
    });
  }

  /**
   * returns an instance of a Bit tester implementation. use components like mocha-tester or
   * jest-tester or [build your own](https://bit.dev/docs/tester/implement-tester).
   */
  tester() {
    // TODO @gilad: check jest tester on watch.
    // TODO: upgrade to jest 29.
    return _defender2().JestTester.from({
      jest: require.resolve('jest'),
      config: this.jestConfigPath
    });
  }

  /**
   * return an instance of a Bit preview instance. used for preview purposes.
   * base preview is a general purpose webpack preview.
   */
  preview() {
    return _preview().ReactPreview.from({
      /**
       * override the default docs template for components.
       */
      // docsTemplate: this.previewDocsTemplate,
      /**
       * mounters are used to mount components to DOM
       * during preview. use them for wrapping your components
       * with routing, theming, data fetching or other types
       * of providers.
       */
      mounter: this.previewMounter
      // transformers: [webpackTransformer],
      // webpackModulePath: require.resolve("webpack"),
      // webpackDevServerModulePath: require.resolve("webpack-dev-server")
    });
  }

  /**
   * returns an instance of the default TypeScript extractor.
   * used by default for type inference for both JS and TS.
   */
  schemaExtractor() {
    return _typescript().TypeScriptExtractor.from({
      tsconfig: this.tsconfigPath
    });
  }

  /**
   * returns an instance of the default ESLint.
   * config files would be used to validate coding standards in components.
   * bit will write the minimum required files in any workspace to optimize
   * for dev experience.
   */
  linter() {
    return _defender().ESLintLinter.from({
      tsconfig: this.tsconfigPath,
      eslint: _eslint().ESLint,
      configPath: this.eslintConfigPath,
      // resolve all plugins from the react environment.
      pluginsPath: __dirname,
      extensions: this.eslintExtensions
    });
  }

  /**
   * used by bit's formatter. helps to standardize and automate
   * code styling. can be used a build task or using bit lint and common
   * ides.
   */
  formatter() {
    return _defender3().PrettierFormatter.from({
      configPath: this.prettierConfigPath
    });
  }

  /**
   * define the build pipeline for a component.
   * pipelines are optimized for performance and consistency.
   * making sure every component is independently built and tested.
   */
  build() {
    // :TODO fix build task
    return _builder().Pipeline.from([_typescript2().TypescriptTask.from({
      tsconfig: this.tsconfigPath,
      // types: resolveTypes(__dirname, ["./types"]),
      typescript: _typescript3().default
    }), _defender().EslintTask.from({
      tsconfig: this.tsconfigPath,
      eslint: _eslint().ESLint,
      configPath: this.eslintConfigPath,
      // resolve all plugins from the react environment.
      pluginsPath: __dirname,
      extensions: this.eslintExtensions
    }), _defender2().JestTask.from({
      config: this.jestConfigPath,
      jest: require.resolve('jest')
    })]);
  }

  /**
   * add build tasks to execute upon [snap](https://bit.dev/docs/snaps).
   * use the snap pipeline for staging and test deployments
   */
  snap() {
    return _builder().Pipeline.from([]);
  }

  /**
   * add build tasks to execute upon [tag](https://bit.dev/docs/tags).
   * use the tag pipeline for deployments, or other tasks required for
   * publishing a semantic version for a component.
   */
  tag() {
    return _builder().Pipeline.from([]);
  }
  workspaceConfig() {
    return _workspaceConfigFiles().ConfigWriterList.from([_typescript2().TypescriptConfigWriter.from({
      tsconfig: this.tsconfigPath
      // types: resolveTypes(__dirname, [this.tsTypesPath]),
    }), _defender().EslintConfigWriter.from({
      configPath: this.eslintConfigPath,
      tsconfig: this.tsconfigPath
    }), _defender3().PrettierConfigWriter.from({
      configPath: this.prettierConfigPath
    })]);
  }

  /**
   * a list of starters for new projects. this helps create a quick and
   * standardized
   */
  starters() {
    return _generator().StarterList.from([_reactGenerator2().ReactWorkspaceStarter.from()]);
  }

  /**
   * set a list of component templates to use across your
   * workspaces. new workspaces would be set to include
   * your envs by default.
   */
  generators() {
    return _generator().TemplateList.from([_reactGenerator().ReactComponentTemplate.from(), _reactGenerator().ReactHookTemplate.from(), _reactGenerator().ReactContextTemplate.from(), _reactGenerator().ReactAppTemplate.from(), _reactGenerator().ReactJSComponentTemplate.from(), _reactGenerator().ReactEnvTemplate.from()]);
  }
  apps() {
    return _application().AppTypeList.from([_reactApps().ReactAppType.from()]);
  }

  /**
   * configure and control the packaging process of components.
   */
  package() {
    return _pkg().PackageGenerator.from({
      packageJson: this.packageJson,
      npmIgnore: this.npmIgnore
    });
  }
}
exports.ReactEnv = ReactEnv;
var _default = exports.default = new ReactEnv();
