// @teambit/ui-foundation.ui.pages.preview-not-found@0.0.80/dist/preview-not-found.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewNotFoundPage = void 0;
const react_1 = __importDefault(require("react"));
const image_icon_1 = require("./image-icon");
const styles = {
    container: {
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        height: '100%',
    },
    image: {
        width: '2.6em',
        marginBottom: '1em',
    },
    message: {
        fontSize: '1em',
        textAlign: 'center',
    },
};
function PreviewNotFoundPage(props) {
    return (react_1.default.createElement("div", Object.assign({}, props, { style: Object.assign(Object.assign({}, styles.container), props.style) }),
        react_1.default.createElement(image_icon_1.ImageIcon, { style: styles.image }),
        react_1.default.createElement("div", { style: styles.message }, "No preview available")));
}
exports.PreviewNotFoundPage = PreviewNotFoundPage;
