// @teambit/design.ui.pages.server-error@0.0.366/dist/server-error-page.docs.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = __importDefault(require("react"));
const documenter_ui_section_1 = require("@teambit/documenter.ui.section");
const documenter_theme_theme_compositions_1 = require("@teambit/documenter.theme.theme-compositions");
const documenter_ui_separator_1 = require("@teambit/documenter.ui.separator");
const server_error_page_1 = require("./server-error-page");
function Overview() {
    return (react_1.default.createElement(documenter_theme_theme_compositions_1.ThemeCompositions, null,
        react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(documenter_ui_section_1.Section, null, "A page component that displays a 500 error message."),
            react_1.default.createElement(documenter_ui_separator_1.Separator, null))));
}
exports.default = Overview;
Overview.abstract = 'A 500 error page.';
Overview.labels = ['react', 'typescript', '500', 'error', 'page'];
Overview.examples = [
    {
        scope: {
            ServerErrorPage: server_error_page_1.ServerErrorPage,
        },
        title: '500 page',
        description: '500 error message',
        jsx: react_1.default.createElement(server_error_page_1.ServerErrorPage, null),
    },
];
