// @teambit/react.apps.react-app-types@1.0.19/dist/web/webpack/webpack.app.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ssrBuildConfig = exports.buildConfig = exports.ssrConfig = exports.clientConfig = exports.calcOutputPath = exports.PUBLIC_PATH = exports.SSR_ENTRY_FILE = void 0;
const path_1 = require("path");
const OUTPUT_FOLDER = 'public/bit-apps';
exports.SSR_ENTRY_FILE = 'index.js';
/** _external_ base url for static files. (is not a folder)
 * @example
 * "http://localhost:3000/public/main.88a88d47d86aaabc0230.js"
 */
exports.PUBLIC_PATH = '/public';
var Targets;
(function (Targets) {
    Targets["ssr"] = "ssr";
    Targets["browser"] = "browser";
})(Targets || (Targets = {}));
function calcOutputPath(appName, target) {
    return (0, path_1.join)(OUTPUT_FOLDER, appName, target);
}
exports.calcOutputPath = calcOutputPath;
function clientConfig() {
    return {
        output: { publicPath: exports.PUBLIC_PATH, filename: '[name].[chunkhash].js' },
    };
}
exports.clientConfig = clientConfig;
function ssrConfig() {
    return {
        target: 'node',
        devtool: 'eval-cheap-source-map',
        output: {
            libraryTarget: 'commonjs',
            filename: exports.SSR_ENTRY_FILE,
        },
    };
}
exports.ssrConfig = ssrConfig;
function buildConfig({ outputPath, }) {
    return {
        output: {
            path: outputPath,
            publicPath: `/`,
            filename: '[name].[chunkhash].js',
        },
    };
}
exports.buildConfig = buildConfig;
function ssrBuildConfig({ outputPath, }) {
    return {
        target: 'node',
        devtool: 'eval-cheap-source-map',
        output: {
            path: outputPath,
            publicPath: `/`,
            libraryTarget: 'commonjs',
            filename: 'index.js',
        },
    };
}
exports.ssrBuildConfig = ssrBuildConfig;
