// @teambit/ui-foundation.ui.pages.static-error@0.0.88/dist/static-error-pages.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.noPreview = exports.serverError = exports.notFound = void 0;
const react_1 = __importDefault(require("react"));
const base_ui_theme_fonts_book_1 = require("@teambit/base-ui.theme.fonts.book");
const design_ui_pages_not_found_1 = require("@teambit/design.ui.pages.not-found");
const design_ui_pages_server_error_1 = require("@teambit/design.ui.pages.server-error");
const ui_foundation_ui_pages_preview_not_found_1 = require("@teambit/ui-foundation.ui.pages.preview-not-found");
const render_page_1 = require("./render-page");
const center = `
  body { 
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;

    font-family: sans-serif;
  }
`;
const sizing = `
  body {
    font-size: 18px;
    color: #878c9a;
    width: 100% !important;
    height: 500px;
  }

  @media screen and (max-width: 250px) {
    body {
      font-size: 14px;
      color: #c7c7c7;
    }
  }
`;
const assets = {
    style: [center],
    css: [base_ui_theme_fonts_book_1.staticBookFontUrl],
};
const noPreviewAssets = {
    style: [center, sizing],
    css: [base_ui_theme_fonts_book_1.staticBookFontUrl],
};
function notFound() {
    return (0, render_page_1.fullPageToStaticString)(react_1.default.createElement(design_ui_pages_not_found_1.NotFoundPage, { className: base_ui_theme_fonts_book_1.staticBookFontClass }), assets);
}
exports.notFound = notFound;
function serverError() {
    return (0, render_page_1.fullPageToStaticString)(react_1.default.createElement(design_ui_pages_server_error_1.ServerErrorPage, { className: base_ui_theme_fonts_book_1.staticBookFontClass }), assets);
}
exports.serverError = serverError;
function noPreview() {
    return (0, render_page_1.fullPageToStaticString)(react_1.default.createElement(ui_foundation_ui_pages_preview_not_found_1.PreviewNotFoundPage, { className: base_ui_theme_fonts_book_1.staticBookFontClass }), noPreviewAssets);
}
exports.noPreview = noPreview;
